SapMachine #820: Bring Vitals up to date

tstuefe opened the pull request:
This is a collection of changes to the Vitals. Changes have accumulated, and its time to bring them to the sapmachine. The biggest changes are in vitals/vitals.cpp, formerly services/stathist.cpp. Review-wise it may be easier just to look at that file in its current form (specifically, all classes starting with "Sample") instead of parsing the diff, but I leave this to you.

Changes in detail:

- consistently named the Vitals Vitals, removing all traces of the former "stathist" or "StatisticalHistory"
- moved the code into an own sub directory (/share/vitals)
- On Linux, added a new column, named "hp" for "process heap", which contains the sum of rss+swap for the process heap (I included swap in case we are swapping and rss is not the full story). The data are read from proc/pid/smaps.
- "productified" most Vitals options, removing the Diagnostic marker.
-  Added a new option to print vitals to stdout at VM exit (we had already "DumpVitalsAtExit", same thing different sink)
- now there are some (basic) gtests for vitals
- and a new jtreg stress test printing vitals in quick succession

vitals.cpp:
- I now use a Mutex to guard table update and table printing.
- Where before the downsample tables were cascaded from the first table (so, samples falling out of the first FIFO were fed into the subsequent mid term table, and falling out of that one into the long term table), now the tables are independent and are fed the same samples, just with different periods. So, the short term table gets every sample, the mid term table every nth sample, etc. The effect is that each table can be looked at individually, without having to know the other tables. And that the time range the downsample tables cover starts now, not at now-x.
- Iterating over the sample tables was streamlined and simplified. Walking the table is done with a closure interface now.

Sorry for the big change. I hope after this change things become reasonably stable.

Testing: I will let the patch cook in our nightlies for a while.

fixes #820

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

RealCLanger commented:
There are gtest errors on Mac related to vitals...

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

RealCLanger commented:
@tstuefe I think you want to merge the current sapmachine brach into your PR as we integrated 18+4 yesterday.

ArnoZeller commented:
Hi Thomas,

there is a build issue because off a warning on windows (and handling warning as errors):
...\sapmachine\src\hotspot\share\vitals\vitals.cpp(162): error C2220: warning treated as error - no 'object' file generated
...\\sapmachine\src\hotspot\share\vitals\vitals.cpp(162): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

RealCLanger reviewed (approved):
Thanks for doing this. I didn't do a detailed technical review but the new naming and code structure looks nice.

schmelter-sap commented on src/hotspot/share/vitals/vitals.cpp:
```diff
@@ -111,19 +143,24 @@ static void print_text_with_dashes(outputStream* st, const char* text, int width
   va_end(args);
-  assert((size_t)len < sizeof(buf), "Truncation. Increase bufsize.");
+  // jio_vsnprintf guarantees -1 on truncation, and always zero termination if buffersize > 0.
+  assert(len != -1, "Truncation. Increase bufsize.");
```
jio_vsnprintf  kann auch eine andere negative Zahl als -1 im Fehlerfall zurückgeben.

schmelter-sap commented on src/hotspot/share/vitals/vitals.cpp:
```diff
@@ -531,462 +594,330 @@ static void print_one_record(outputStream* st, const record_t* record,
+    st->cr();
 
-  }
+    st->print_raw_cr("Last 60 minutes:");
```
Die Zeitangaben (60 min, 24 h, 10 d) stimmen nur  wenn VitalsSampleInterval 10 Sekunden ist.

schmelter-sap reviewed (approved):
Sieht gut aus. Die beiden Anmerkungen sind nicht kriegsentscheidend.

tstuefe commented:
Thanks Ralf!

I fixed the time span display in a way that works if the sample interval is different. Also added a test fix for non-linux platforms (not every platform has the --system-- columns).

..Thomas

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

schmelter-sap reviewed (approved):
LGTM

SapMachine commented:
Hello @tstuefe, this pull request fulfills all formal requirements.

RealCLanger commented:
I merged current sapmachine into this branch, testing quality should be better.
